### ISSUE TITLE
sync with --cf-invalidate gives "S3 error: 403 (AccessDenied)"

### ISSUE THREAD

**Issue by jpbernius** [s3tools/s3cmd#750]
Hello,

first of all many thanks for developing such a great tool.

I keep getting the Access Denied and I strongly believe it is the CloudFront step.

**Command**:

```
s3cmd sync \
    --acl-public \
    --delete-removed \
    --no-mime-magic \
    --cf-invalidate \
    --access_key=$AWS_ACCESS_KEY_ID \
    --secret_key=$AWS_SECRET_ACCESS_KEY \
    --verbose \
    ./build/ \
    $AWS_BUCKET_URL_PRODUCTION
```

From **Log**:

```
INFO: Compiling list of local files...
INFO: Running stat() and reading/calculating MD5 values on 160 files, this may take some time...
INFO: Retrieving list of remote files for s3://janphilip.bernius.net/ ...
INFO: Forwarding request to eu-central-1
INFO: Found 160 local files, 158 remote files
INFO: Verifying attributes...
INFO: Summary: 7 local files to upload, 0 files to remote copy, 0 remote files to delete
upload: './build/javascripts/lang_redirect_blog-d973ae56.js' -> 's3://janphilip.bernius.net/javascripts/lang_redirect_blog-d973ae56.js'  [1 of 7]
 255 of 255   100% in    0s  1049.42 B/s  done
upload: './build/javascripts/lang_redirect_home-57e15325.js' -> 's3://janphilip.bernius.net/javascripts/lang_redirect_home-57e15325.js'  [2 of 7]
 245 of 245   100% in    0s  1106.01 B/s  done
upload: './build/blog/index.html' -> 's3://janphilip.bernius.net/blog/index.html'  [3 of 7]
 382 of 382   100% in    0s  1637.05 B/s  done
upload: './build/de/index.html' -> 's3://janphilip.bernius.net/de/index.html'  [4 of 7]
 18402 of 18402   100% in    0s    81.03 kB/s  done
upload: './build/en/index.html' -> 's3://janphilip.bernius.net/en/index.html'  [5 of 7]
 19343 of 19343   100% in    0s    91.85 kB/s  done
upload: './build/index.html' -> 's3://janphilip.bernius.net/index.html'  [6 of 7]
 364 of 364   100% in    0s  1410.87 B/s  done
upload: './build/sitemap.xml' -> 's3://janphilip.bernius.net/sitemap.xml'  [7 of 7]
 1852 of 1852   100% in    0s     7.60 kB/s  done
Done. Uploaded 40843 bytes in 1.6 seconds, 24.47 kB/s.
ERROR: S3 error: 403 (AccessDenied): Access Denied
```

IAM **Policy**:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:ListAllMyBuckets",
                "s3:GetBucketWebsite"
            ],
            "Effect": "Allow",
            "Resource": "arn:aws:s3:::*"
        },
        {
            "Action": [
                "s3:DeleteObject",
                "s3:GetObject",
                "s3:GetObjectAcl",
                "s3:ListBucket",
                "s3:PutObject",
                "s3:PutObjectAcl"
            ],
            "Effect": "Allow",
            "Resource": [
                "arn:aws:s3:::janphilip.bernius.net",
                "arn:aws:s3:::janphilip.bernius.net/*"
            ]
        },
        {
            "Action": [
                "cloudfront:CreateInvalidation"
            ],
            "Effect": "Allow",
            "Resource": [
                "*"
            ]
        }
    ]
}
```

I am new to AWS, so please forgive me if this is an obvious mistake.  
Thanks,
Jan Philip

**Comment by fviard** [comment]
Hi Jan,
I'm not expert in cloudfront, so I don't know yet what policy you could be missing, but you should run s3cmd with the "-d" flag instead of the "--verbose". That will gives us the complete debug log with the exact request that is failing.

**Comment by fviard** [comment]
Hi @jpbernius, were you able to fix your issue or get a debug log?

**Comment by fviard** [comment]
Close as no news. Please re-open if you still encounter the issue and can provide more details.

**Comment by ramynasr** [comment]
I am having exactly the same issue using s3cmd 1.6.0

```
DEBUG: Canonical Request:
GET
/
website=
host:someHostNameRemovedHere.s3.amazonaws.com
x-amz-content-sha256:someSha256RemovedHere
x-amz-date:20170131T154533Z

host;x-amz-content-sha256;x-amz-date
someSha256RemovedHere
----------------------
DEBUG: signature-v4 headers: {'x-amz-content-sha256': 'someSha256RemovedHere', 'Authorization': u'AWS4-HMAC-SHA256 Credential=someKeyRemovedHere/20170131/us-east-1/s3/aws4_request,SignedHeaders=host;x-amz-content-sha256;x-amz-date,Signature=someSignatureRemovedHere', 'x-amz-date': '20170131T154533Z'}
DEBUG: Processing request, please wait...
DEBUG: get_hostname(someHostNameRemovedHere): someHostNameRemovedHere.s3.amazonaws.com
DEBUG: ConnMan.get(): re-using connection: http://someHostNameRemovedHere.s3.amazonaws.com#87
DEBUG: format_uri(): /?website
DEBUG: Sending request method_string='GET', uri='/?website', headers={'x-amz-content-sha256': 'someSha256RemovedHere', 'Authorization': u'AWS4-HMAC-SHA256 Credential=someKeyRemovedHere/20170131/us-east-1/s3/aws4_request,SignedHeaders=host;x-amz-content-sha256;x-amz-date,Signature=someSignatureRemovedHere', 'x-amz-date': '20170131T154533Z'}, body=(0 bytes)
DEBUG: Response: {'status': 403, 'headers': {'x-amz-id-2': 'someIdRemovedHere', 'server': 'AmazonS3', 'transfer-encoding': 'chunked', 'x-amz-request-id': 'someRequestIdRemovedHere', 'date': 'Tue, 31 Jan 2017 15:45:32 GMT', 'content-type': 'application/xml'}, 'reason': 'Forbidden', 'data': '<?xml version="1.0" encoding="UTF-8"?>\n<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>someRequestIdRemovedHere</RequestId><HostId>someIdRemovedHere</HostId></Error>'}
DEBUG: ConnMan.put(): connection put back to pool (http://someHostNameRemovedHere.s3.amazonaws.com#88)
DEBUG: S3Error: 403 (Forbidden)
DEBUG: HttpHeader: x-amz-id-2: someIdRemovedHere
DEBUG: HttpHeader: server: AmazonS3
DEBUG: HttpHeader: transfer-encoding: chunked
DEBUG: HttpHeader: x-amz-request-id: someRequestIdRemovedHere
DEBUG: HttpHeader: date: Tue, 31 Jan 2017 15:45:32 GMT
DEBUG: HttpHeader: content-type: application/xml
DEBUG: ErrorXML: Code: 'AccessDenied'
DEBUG: ErrorXML: Message: 'Access Denied'
DEBUG: ErrorXML: RequestId: 'someRequestIdRemovedHere'
DEBUG: ErrorXML: HostId: 'someIdRemovedHere'
ERROR: S3 error: 403 (AccessDenied): Access Denied
```

**Comment by kichik** [comment]
I was also seeing this issue. I'm not sure which permissions I'm missing for the `/?website` call, but I was able to get rid of it by passing `--cf-no-invalidate-default-index-root` on the command line. After that, it failed the same way on `/?location` so I passed `--region us-east-1` and now it finally works.

I should add that this issue only manifested on my build system. It worked perfectly fine on my development machine, even though I made sure they both use the same IAM user. My `.aws` configuration does specify a default region, so I can see how `/?location` call was only made on my build server. I'm still unsure why the `/?website` call was never made locally.

One more permission I had to add to my IAM user along the way was `cloudfront:ListDistributions`. This too was only required on my build server for an unknown reason.
